### PR TITLE
Add helpers to serialize/deserialize e2e test framework objects

### PR DESCRIPTION
**Pull request by pzread**
This is part of work to migrate benchmark tools (#11076) to the e2e test framework.

With the serialization, we can export the selected `E2EModelRunConfig` for each target device and feed to the benchmark tool to run the benchmarks.

The serialized configs also allow user to modify the benchmark configs when needed.

### CODE REVIEW

File: build_tools/python/e2e_test_framework/serialization.py [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+                                        keyed_obj_map, obj_cache)
+      derialized_obj = cls(**field_value_map)
+      obj_cache[obj_key] = derialized_obj
+      return derialized_obj
+
+    setattr(cls, SERIALIZE_FUNC_NAME, serialize)
+    setattr(cls, DESERIALIZE_FUNC_NAME, deserialize)
+    return cls
+
+  if cls is None:
```
Comment (GMNGeoffrey) [inline build_tools/python/e2e_test_framework/serialization.py:245]: ? That seems like an odd condition

File: build_tools/python/e2e_test_framework/serialization.py [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+    if not dataclasses.is_dataclass(cls):
+      raise ValueError(f"{cls} is not a dataclass.")
+
+    fields = dataclasses.fields(cls)
+
+    def serialize(self, keyed_obj_map: OrderedDict[str, Any]):
+      if not keyed_obj:
+        return _fields_to_dict(self, fields, keyed_obj_map)
+
+      obj_id = getattr(self, id_field)
```
Comment (GMNGeoffrey) [inline build_tools/python/e2e_test_framework/serialization.py:203]: Is it possible to validate in the main body of the function that the class has such a field. Also could check that it's a dataclass. I think this doesn't need to be deferred.

File: build_tools/python/e2e_test_framework/serialization_test.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+  TEST_OBJS = [
+      TestA(b_list=[B_OBJ_A, B_OBJ_B],
+            c_obj=TestC(float_val=0.1),
+            str_val="test1",
+            enum_val=EnumX.OPTION_B),
+      TestA(b_list=[B_OBJ_A],
+            c_obj=TestC(float_val=0.2),
+            str_val=None,
+            enum_val=EnumX.OPTION_C)
+  ]
```
Comment (GMNGeoffrey) [inline build_tools/python/e2e_test_framework/serialization_test.py:72]: I'd recommend against moving these into class fields. Each test function should stand on its own and this hides relevant properties of the test

File: build_tools/python/e2e_test_framework/serialization_test.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+  def test_roundtrip_with_json(self):
+    objs = collections.OrderedDict(
+        x=self.B_OBJ_A,
+        y=self.B_OBJ_B,
+    )
+
+    json_str = json.dumps(serialization.serialize_and_pack(objs))
+    results = serialization.unpack_and_deserialize(
+        json.loads(json_str), typing.OrderedDict[str, TestB])
```
[the remark this answers is not in the document]
Comment (pzread) [reply build_tools/python/e2e_test_framework/serialization_test.py:136]: JSON custom encoder/decoder isn't very suitable in this case due to the way it traverses the objects.

In JSON custom encoder you implmenet the `default(obj)` method to handle the unknown objects. The way it works is that your `default(obj)` serializes only one level of the object and returns it. For example, given a dataclass object:
```
@dataclass
class ObjA(object):
  id: str
  value: ObjB
```
You will return
```
{
  id: "...",
  value: object_b
}
```
Then JSON encoder will recursively traverse your returned object (a dict in this case) and call your `default(obj)` again for unknown objects (`object_b` in this case).

The problem here is for keyed objects, we want to replace them with their IDs and store the serialized objects in a separate map. The way JSON encoder works makes it hard to do that. If you return an ID, JSON encoder won't try to serialize the keyed object, so you'll need to track those keyed objects and serialize them again in post-processing.

Another problem is that we want to compose the `root_obj` and `keyed_obj_map` as the final serialization result. But JSON custom encoder also doesn't provide an option to do such post-composition (and it's hard to merge two serialized JSON strings. It's easier to compose two values into a dict first then run `json.dumps()` one it)

File: build_tools/python/e2e_test_framework/serialization_test.py (hunk 3 above)
Comment (pzread) [reply build_tools/python/e2e_test_framework/serialization_test.py]: Done.

File: build_tools/python/e2e_test_framework/serialization.py (hunk 2 above)
Comment (pzread) [reply build_tools/python/e2e_test_framework/serialization.py:203]: Done.

File: build_tools/python/e2e_test_framework/serialization.py (hunk 1 above)
Comment (pzread) [reply build_tools/python/e2e_test_framework/serialization.py:242]: This is the trick to make the uses of both `@serializable` and `@serializable(...)` possible. Documented.

File: build_tools/python/e2e_test_framework/serialization.py [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+    type_key: string defining the object type, must be set if keyed_obj is True.
+    id_field: field name of the id field of a keyed object.
+
+  Example:
+    @serializable
+    @dataclass
+    class A(object):
+      ...
+
+    @serialzable(keyed_obj=True, type_key="obj_b")
```
[the remark this answers is not in the document]
Comment (pzread) [reply build_tools/python/e2e_test_framework/serialization.py:182]: Done.

File: build_tools/python/e2e_test_framework/serialization.py (hunk 1 above)
Comment (GMNGeoffrey) [reply build_tools/python/e2e_test_framework/serialization.py:244]: Aaaah I get it

**Review by GMNGeoffrey — APPROVED** [review]

**Pull request merged**